crypto/rand/rand_unix.c

```diff
@@ -119,26 +134,89 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 #   error "Seeding uses urandom but DEVRANDOM is not configured"
 #  endif
 
+#  if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#   if __GLIBC_PREREQ(2, 25)
+#    define OPENSSL_HAVE_GETRANDOM
+#   endif
+#  endif
+
+#  if (defined(__FreeBSD__) && __FreeBSD_version >= 1200061)
+#   define OPENSSL_HAVE_GETRANDOM
+#  endif
+
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+#   include <sys/random.h>
+#  endif
+
 #  if defined(OPENSSL_RAND_SEED_OS)
 #   if !defined(DEVRANDOM)
 #    error "OS seeding requires DEVRANDOM to be configured"
 #   endif
+#   define OPENSSL_RAND_SEED_GETRANDOM
 #   define OPENSSL_RAND_SEED_DEVRANDOM
-#   if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
-#    if __GLIBC_PREREQ(2, 25)
-#     define OPENSSL_RAND_SEED_GETRANDOM
-#    endif
-#   endif
-#  endif
-
-#  ifdef OPENSSL_RAND_SEED_GETRANDOM
-#   include <sys/random.h>
 #  endif
 
 #  if defined(OPENSSL_RAND_SEED_LIBRANDOM)
 #   error "librandom not (yet) supported"
 #  endif
 
+#  if defined(__FreeBSD__) && defined(KERN_ARND)
+/*
+ * sysctl_random(): Use sysctl() to read a random number from the kernel
+ * Returns the size on success, 0 on failure.
+ */
+static size_t sysctl_random(void *buf, size_t buflen)
+{
+    int mib[2];
+    size_t done = 0;
+    size_t len;
+
+    /* The sysctl returns longs. */
+    if (buflen % sizeof(long) != 0)
+        return 0;
+
+    mib[0] = CTL_KERN;
+    mib[1] = KERN_ARND;
+
+    do {
+        len = buflen;
+        if (sysctl(mib, 2, buf, &len, NULL, 0) == -1)
+            return done;
+        done += len;
+        buf += len;
+        buflen -= len;
+    } while (buflen > 0);
+
+    return done;
+}
+#  endif
+
+/*
+ * syscall_random(): Try to get random data using a system call
+ * returns the number of bytes returned in buf, or <= 0 on error.
+ */
+int syscall_random(void *buf, size_t buflen)
+{
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+    return (int)getrandom(buf, buflen, 0);
+#  endif
+
+#  if defined(__linux) && defined(SYS_getrandom)
+    return (int)syscall(SYS_getrandom, buf, buflen, 0);
```

Comment:
I'm only a layman in this matter, so please excuse me if my question is off-topic or stupid: In this [LWN Article](https://lwn.net/Articles/711013) the author Jonathan Corbet describes that there was a lot of discussion about whether the glibc getrandom() call should be a thread cancellation point or not. Finally it was decided that it should be.

```
The most extensive argument, though, was over whether getrandom() should be a thread cancellation point.
In other words, what should happen if pthread_cancel() is called on a thread that is currently blocked
in getrandom()? The original patch did make getrandom() into a cancellation point; it still behaves
that way in the version merged for 2.25, but it had to survive a lot of argument to get there.
```

While the glibc getrandom() is a thread cancellation point, your raw syscall is not. Does this difference have any practical relevance for OpenSSL?

Reply:
I don't think it's relevant. Cancellation points are just a bunch of functions where the cancellation can happen, which are typically function that can block.